google/cloud/storage/internal/grpc_client.cc

```diff
@@ -286,28 +297,40 @@ StatusOr<BucketMetadata> GrpcClient::LockBucketRetentionPolicy(
 
 StatusOr<ObjectMetadata> GrpcClient::InsertObjectMedia(
     InsertObjectMediaRequest const& request) {
-  grpc::ClientContext context;
-  auto auth = auth_strategy_->ConfigureContext(context);
-  if (!auth.ok()) return auth;
-  google::storage::v1::Object response;
-  auto stream = stub_->InsertObject(&context, &response);
+  auto stream = [&]() -> std::unique_ptr<InsertStream> {
+    using InsertStreamImpl = google::cloud::internal::StreamingWriteRpcImpl<
+        google::storage::v1::InsertObjectRequest, google::storage::v1::Object>;
+    using InsertStreamError = google::cloud::internal::StreamingWriteRpcError<
+        google::storage::v1::InsertObjectRequest, google::storage::v1::Object>;
+
+    auto context = absl::make_unique<grpc::ClientContext>();
+    auto auth = auth_strategy_->ConfigureContext(*context);
+    if (!auth.ok()) {
+      return absl::make_unique<InsertStreamError>(std::move(auth));
+    }
+    auto holder = absl::make_unique<google::storage::v1::Object>();
+    auto impl = stub_->InsertObject(context.get(), holder.get());
+    return absl::make_unique<InsertStreamImpl>(
+        std::move(context), std::move(holder), std::move(impl));
+  }();
+
   auto proto_request = ToProto(request);
-  std::size_t const maximum_buffer_size =
-      google::storage::v1::ServiceConstants::MAX_WRITE_CHUNK_BYTES;
   auto const& contents = request.contents();
+  auto const contents_size = static_cast<std::int64_t>(contents.size());
+  std::int64_t const maximum_buffer_size =
+      google::storage::v1::ServiceConstants::MAX_WRITE_CHUNK_BYTES;
 
   // This loop must run at least once because we need to send at least one
   // Write() call for empty objects.
-  std::size_t offset = 0;
-  do {
+  for (std::int64_t offset = 0, n = 0; offset < contents_size; offset += n) {
```

Comment:
Since this loop must run at least once, is `contents_size` guaranteed to be > 0?

Reply:
Nope, thanks for catching that. Fixed now.